docs/compilers/CSharp/Runtime Async Design.md

```diff
@@ -0,0 +1,543 @@
+    IL_0005: callvirt instance valuetype [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter [System.Runtime]System.Threading.Tasks.Task::GetAwaiter()
+    IL_000a: stloc.0
+    IL_000b: ldloca.s 0
+    IL_000d: call instance bool [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter::get_IsCompleted()
+    IL_0012: brtrue.s IL_001b
+
+    IL_0014: ldloc.0
+    IL_0015: call class [System.Runtime]System.Threading.Tasks.Task System.Runtime.CompilerServices.RuntimeHelpers::AwaitAwaiterFromRuntimeAsync<valuetype [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter>(!!0)
+    IL_001a: pop
+
+    IL_001b: ldloca.s 0
+    IL_001d: call instance void [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter::GetResult()
+    IL_0022: ret
+}
+```
+
+#### Await local of concrete type `Task<T>`
+
+This strategy will also be used for `Task`-like types that are not `Task`, `ValueTask`, `Task<T>`, or `ValueTask<T>`, both in value form, and in direct method call form. We will use either `AwaitAwaiterFromRuntimeAsync` or
+`UnsafeAwaitAwaiterFromRuntimeAsync`, depending on the interface implemented by the custom awaitable.
+
+```cs
+var local = M();
+var i = await local;
+
+class C
+{
+    static Task<int> M();
+}
+```
+
+Translated C#:
+
+```cs
+var local = C.M();
+var i =
+{
+    var awaiter = local.GetAwaiter();
+    if (!awaiter.IsComplete)
+    {
+        /* Runtime-Async Call */ System.Runtime.CompilerServices.RuntimeHelpers.AwaitAwaiterFromRuntimeAsync<System.Runtime.CompilerServices.TaskAwaiter>(awaiter);
+    }
+    awaiter.GetResult()
+};
+```
+
+```il
+{
+    .locals init (
+        [0] valuetype [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter`1<int32> awaiter
+    )
+
+    IL_0000: call class [System.Runtime]System.Threading.Tasks.Task`1<int32> C::M()
+    IL_0005: callvirt instance valuetype [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter`1<!0> class [System.Runtime]System.Threading.Tasks.Task`1<int32>::GetAwaiter()
+    IL_000a: stloc.0
+    IL_000b: ldloca.s 0
+    IL_000d: call instance bool valuetype [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter`1<int32>::get_IsCompleted()
+    IL_0012: brtrue.s IL_001b
+
+    IL_0014: ldloc.0
+    IL_0015: call class [System.Runtime]System.Threading.Tasks.Task System.Runtime.CompilerServices.RuntimeHelpers::AwaitAwaiterFromRuntimeAsync<valuetype [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter`1<int32>>(!!0)
+    IL_001a: pop
+
+    IL_001b: ldloca.s 0
+    IL_001d: call instance !0 valuetype [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter`1<int32>::GetResult()
+    IL_0022: pop
+    IL_0023: ret
+}
+```
+
+#### Await a `T`-returning method
+
+```cs
+await C.M<Task>();
+
+class C
+{
+    static T M<T>();
+}
+```
+
+```il
+TODO: https://github.com/dotnet/runtime/issues/109632
+```
+
+#### Await a generic `T` `Task<T>`-returning method
+
+```cs
+await C.M<int>();
+
+class C
+{
+    static Task<T> M<T>();
+}
```

Comment:
This is just a regular case. Substituting the element type is not controversial.   (unlike substituting the entire return type)